ui/dashboards/src/components/PanelDrawer/PanelSpecEditor.tsx

```diff
@@ -0,0 +1,73 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { ErrorAlert, JSONEditor } from '@perses-dev/components';
+import { PanelDefinition, QueryDefinition, UnknownSpec } from '@perses-dev/core';
+import {
+  usePlugin,
+  PanelPlugin,
+  OptionsEditorTabsProps,
+  TimeSeriesQueryEditor,
+  OptionsEditorTabs,
+} from '@perses-dev/plugin-system';
+
+export interface PanelSpecEditorProps {
+  panelDefinition: PanelDefinition;
+  onQueriesChange: (queries: QueryDefinition[]) => void;
+  onPluginSpecChange: (spec: UnknownSpec) => void;
+  onJSONChange: (panelDefinition: PanelDefinition) => void;
+}
+
+export function PanelSpecEditor(props: PanelSpecEditorProps) {
+  const { panelDefinition, onJSONChange, onQueriesChange, onPluginSpecChange } = props;
+  const { kind } = panelDefinition.spec.plugin;
+  const { data: plugin, isLoading, error } = usePlugin('Panel', kind);
+
+  if (error) {
+    return <ErrorAlert error={error} />;
+  }
+
+  // TODO: Proper loading indicator
+  if (isLoading) {
+    return null;
+  }
+
+  if (plugin === undefined) {
+    throw new Error(`Missing implementation for panel plugin with kind '${kind}'`);
+  }
+
+  const { panelOptionsEditorComponents, hideQueryEditor } = plugin as PanelPlugin;
+  let tabs: OptionsEditorTabsProps['tabs'] = [];
+
+  if (!hideQueryEditor && panelDefinition.spec.queries !== undefined) {
+    // Since we only support TimeSeriesQuery for now, we will always show a TimeSeriesQueryEditor
+    tabs.push({
+      label: 'Query',
+      content: <TimeSeriesQueryEditor queries={panelDefinition.spec.queries} onChange={onQueriesChange} />,
```

Comment:
Decided to add `PanelSpecEditor` instead modifying `PluginSpecEditor` directly because otherwise there is a circular dependency between `PluginSpecEditor` and `TimeSeriesQueryEditor`

Reply:
I've got a handful of questions on this bit to make sure I'm following how all these pieces fit together...

With this change, does `PluginSpecEditor` handle non-panel plugin editor rendering and `PanelSpecEditor` handles panel plugin editor rendering or is it more complex than that?

Would it makes sense to put `PanelSpecEditor` in `plugin-system` where the other plugin rendering components live?

Is there a potential future where we need to support non-timeseries queries? If so, how will that fit with how this is currently being organized? I'm guessing we could add an additional value in the spec that we could use to conditionally render the right thing here, but was curious how you're thinking about it.

Reply:
all valid questions!! 

> With this change, does `PluginSpecEditor` handle non-panel plugin editor rendering and `PanelSpecEditor` handles panel plugin editor rendering or is it more complex than that?

I think it makes sense for `PluginSpecEditor` to handle non-panel plugins while `PanelSpecEditor` handles only panel plugin because `PanelPlugin` type is already different from other plugin types. For instance `PanelPlugin` requires `panelOptionsEditorComponents` which is an array of react components with labels where as other plugins takes in `OptionsEditorComponent`.

However, for now i only removed the part of the code in `PluginSpecEditor` that introduced the circular dependency because I wasn't sure if I was missing some custom panel plugin use case.

This is definitely open for discussion!

>  Would it makes sense to put PanelSpecEditor in plugin-system where the other plugin rendering components live?

Yeah i can move it to plugin-system. The reason why i put it in dashboards package was because I wasn't sure if its confusing to have both PanelSpecEditor and PluginSpecEditor. I think it'll be more clear once we know the answer to the question above!

> Is there a potential future where we need to support non-timeseries queries? If so, how will that fit with how this is currently being organized? I'm guessing we could add an additional value in the spec that we could use to conditionally render the right thing here, but was curious how you're thinking about it.

I think eventually we need to update TimeSeriesQueryEditor to be a generic QueryEditor that supports other metrics like what Shan said in another comment. Also since the framework code knows what queries are being run, i think we can conditionally render the different query editors based on the query type.

Reply:
I moved `PanelSpecEditor` back to plugin-system because it doesn't really make sense to have it inside dashboards/components/PanelDrawer now that i think about it more...